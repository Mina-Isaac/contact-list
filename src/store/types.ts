export interface NameDetails {
  title: string;
  first: string;
  last: string;
};

interface Location {
  street: { number: number; name: string };
  city: string;
  state: string;
  country: string;
  postcode: string;
}

export interface Contact {
  gender: string;
  name: NameDetails;
  location: Location;
  email: string;
  login: { username: string };
  phone: string;
  cell: string;
  picture: {
    large: string;
    medium: string;
    thumbnail: string;
  };
}

export type APIResponse = {
  results: Contact[];
};
