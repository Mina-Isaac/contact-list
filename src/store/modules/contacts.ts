import {
  createAsyncThunk,
  createReducer,
  SerializedError,
} from "@reduxjs/toolkit";
import { AppConfig } from "../../constants";
import { appService } from "../../service";
import { Contact } from "../types";

interface ContactsState {
  loading: "idle" | "pending" | "fulfilled" | "rejected";
  list: Contact[];
  error: SerializedError | null;
}

const initialState: ContactsState = {
  loading: "idle",
  list: [],
  error: null,
};

export const fetchContacts = createAsyncThunk(
  "contacts/fetchData",
  async (config: AppConfig) => {
    const response = await appService.fetchContacts(config);
    return response.results;
  }
);

const contactsReducer = createReducer(initialState, (builder) => {
  builder
    .addCase(fetchContacts.pending, (state) => {
      state.loading = "pending";
    })
    .addCase(fetchContacts.fulfilled, (state, action) => {
      state.list = action.payload;
      state.loading = "fulfilled";
    })
    .addCase(fetchContacts.rejected, (state, action) => {
      state.loading = "rejected";
      state.error = action.error;
    });
});

export default contactsReducer;
