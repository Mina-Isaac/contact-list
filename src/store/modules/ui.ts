import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Contact } from "../types";

interface UIState {
  selectedTab: string;
  selectedContact: Contact | undefined;
}

const initialState: UIState = {
  selectedTab: "a",
  selectedContact: undefined,
};

export const uiSlice = createSlice({
  name: "ui",
  initialState,
  reducers: {
    setSelectedTab: (state, action: PayloadAction<UIState["selectedTab"]>) => {
      state.selectedTab = action.payload;
    },
    setSelectedContact: (
      state,
      action: PayloadAction<UIState["selectedContact"]>
    ) => {
      state.selectedContact = action.payload;
    },
  },
});

export const { setSelectedTab, setSelectedContact } = uiSlice.actions;

export default uiSlice.reducer;
