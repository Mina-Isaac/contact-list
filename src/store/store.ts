import {
  configureStore,
  ThunkAction,
  Action,
  combineReducers,
} from "@reduxjs/toolkit";
import contactsReducer from "./modules/contacts";
import uiReducer from "./modules/ui";

export const rootReducer = combineReducers({
  contacts: contactsReducer,
  ui: uiReducer,
});

export const store = configureStore({
  reducer: rootReducer,
  devTools: process.env.NODE_ENV === "development",
});

export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
