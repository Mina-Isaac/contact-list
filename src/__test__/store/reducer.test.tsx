import { RootState, rootReducer } from "../../store/store";
import { fetchContacts } from "../../store/modules/contacts";
import { setSelectedContact, setSelectedTab } from "../../store/modules/ui";
import { sampleContactList } from "../fixtures/sampleContactList";
import { configJson } from "../../constants";

function getRandomInt(min: number, max: number) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

const getInitialState = (initial?: RootState) =>
  rootReducer(initial as RootState, {} as any);

describe("Reducer data flow", () => {
  it("should match a snapshot", () => {
    const initialState = getInitialState();
    expect(initialState).toMatchSnapshot();
  });

  it("store data on contacts slice on fetchContacts.fulfilled", () => {
    const initialState = getInitialState();
    const state = rootReducer(
      initialState,
      fetchContacts.fulfilled(sampleContactList, "", configJson)
    );
    expect(state.contacts.list).toBe(sampleContactList);
    expect(state.contacts.error).toEqual(null);
    expect(state.contacts.loading).toEqual("fulfilled");
  });

  it("sets selected tab", () => {
    const randomIndex = getRandomInt(0, 10000) % configJson.tabs.length;
    const initialState = getInitialState();
    const randomTab = configJson.tabs[randomIndex];
    const state = rootReducer(initialState, setSelectedTab(randomTab));
    expect(state.ui.selectedTab).toEqual(randomTab);
  });
  it("sets selected contact", () => {
    const initialState = getInitialState();
    const state = rootReducer(
      initialState,
      setSelectedContact(sampleContactList[0])
    );
    expect(state.ui.selectedContact).toBe(sampleContactList[0]);
  });
});
