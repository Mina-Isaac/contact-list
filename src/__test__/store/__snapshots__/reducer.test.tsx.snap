// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer data flow should match a snapshot 1`] = `
Object {
  "contacts": Object {
    "error": null,
    "list": Array [],
    "loading": "idle",
  },
  "ui": Object {
    "selectedContact": undefined,
    "selectedTab": "a",
  },
}
`;
